e2e-tests/testdata/echo_client/ts/client.ts

```diff
@@ -770,6 +770,223 @@ function mustBeSet<A>(field: string, value: A | null | undefined): A {
     return value
 }
 
+function encodeWebSocketHeaders(headers: Record<string, string>) {
+    // url safe, no pad
+    const base64encoded = btoa(JSON.stringify(headers))
+      .replaceAll("=", "")
+      .replaceAll("+", "-")
+      .replaceAll("/", "_");
+    return "encore.dev.headers." + base64encoded;
+}
+
+class WebSocketConnection {
+    public ws: WebSocket;
+    public done = false;
+
+    private msgHandler?: (event: any) => void;
+
+    private url: string;
+    private protocols: string[];
+
+    private retry: number = 0;
+    private minDelayMs: number = 10;
+    private maxDelayMs: number = 500;
+
+    private hasUpdateHandlers: (() => void)[] = [];
+
+    constructor(url: string, headers?: Record<string, string>) {
+        let protocols = ["encore-ws"];
+        if (headers) {
+            protocols.push(encodeWebSocketHeaders(headers))
+        }
+
+        this.protocols = protocols;
+        this.url = url;
+
+        this.ws = this.connect();
+    }
+
+    connect(): WebSocket {
+        const ws = new WebSocket(this.url, this.protocols)
+
+        ws.addEventListener("open", (_event: any) => {
+            this.retry = 0;
+        });
+
+        ws.addEventListener("error", (event: any) => {
+          console.error(event.error);
+          this.ws.close(1002);
+        });
+
+        ws.addEventListener("message", (event: any) => {
+            if (this.msgHandler !== undefined) {
+                this.msgHandler(event);
+            }
+            this.resolveHasUpdateHandlers();
+        });
+
+        ws.addEventListener("close", (event: any) => {
+            // normal closure, no reconnect
+            if (event.code === 1005 || event.code === 1000) {
+                this.done = true;
+            }
+            if (!this.done) {
+                const delay = Math.min(this.minDelayMs * 2 ** this.retry, this.maxDelayMs);
+                console.log(`Reconnecting to ${this.url} in ${delay}ms`);
+                this.retry += 1;
+                setTimeout(() => {
+                    this.ws = this.connect();
```

Comment:
Is automatic reconnect desirable?

Reply:
In my mind it is the expected behavior. Not that this only reconnects on un-graceful terminations.

Otherwise if there is a deploy or a network issue, the connection will be lost. Alternatively we could ofc make it optional or make it possible to setup via some callbacks

Reply:
How does a reconnect behave with a handshake? In my experience some streaming APIs are stateful and indicate an action being taken, and starting a new stream would amount to resubmitting the action, which seems undesirable.

I think it might be better to leave out auto-reconnect for now. If/when we have the ability to tag an endpoint as idempotent it would make more sense.

Reply:
That makes sense, I'll leave it out for now, and maybe add some callbacks so that the caller could potentially implement it themselves if needed (I think I just need to add a "close"-callback to the `on` api)